src/rt-app.c

```diff
@@ -897,175 +897,227 @@ static void set_thread_membind(thread_data_t *data, numaset_data_t * numa_data)
-					perror("setpriority");
-					exit(EXIT_FAILURE);
-				}
-			}
-
-			data->lock_pages = 0; /* forced off for SCHED_OTHER */
+			_set_thread_cfs(data, sched_data);
+			_set_thread_uclamp(data, sched_data);
+			data->lock_pages = 0; /* forced off */
 			break;
-
 		case deadline:
-			log_debug("[%d] setting scheduler %s exec %lu, deadline %lu"
-					" period %lu", data->ind,
-					policy_to_string(sched_data->policy), sched_data->period,
-					sched_data->runtime, sched_data->deadline);
-			log_ftrace(ft_data.marker_fd, FTRACE_ATTRS,
-				   "rtapp_attrs: event=policy policy=dl runtime=%lu deadline=%lu period=%lu",
-				   sched_data->prio, sched_data->runtime,
-				   sched_data->deadline, sched_data->period);
-
-			tid = gettid();
-			sa_params.size = sizeof(struct sched_attr);
-			sa_params.sched_flags = 0;
-			sa_params.sched_policy = SCHED_DEADLINE;
-			sa_params.sched_priority = 0;
-			sa_params.sched_runtime = sched_data->runtime;
-			sa_params.sched_deadline = sched_data->deadline;
-			sa_params.sched_period = sched_data->period;
-
-			ret = sched_setattr(tid, &sa_params, flags);
-			if (ret != 0) {
-				log_critical("[%d] sched_setattr "
-						"returned %d", data->ind, ret);
-				perror("sched_setattr: failed to set deadline attributes");
-				exit(EXIT_FAILURE);
-			}
-		break;
-
+			_set_thread_deadline(data, sched_data);
+			break;
 		default:
 			log_error("Unknown scheduling policy %d",
-					sched_data->policy);
-
+				  sched_data->policy);
 			exit(EXIT_FAILURE);
 	}
 
-	if (sched_data->policy != deadline &&
-	    (sched_data->util_min != -1 || sched_data->util_max != -1)) {
-		sa_params.size = sizeof(struct sched_attr);
-		sa_params.sched_flags = SCHED_FLAG_KEEP_ALL;
-		tid = gettid();
-
-		if (sched_data->util_min != -1) {
-			sa_params.sched_util_min = sched_data->util_min;
-			sa_params.sched_flags |= SCHED_FLAG_UTIL_CLAMP_MIN;
-			log_notice("[%d] setting util_min=%d",
-				   data->ind, sched_data->util_min);
-			log_ftrace(ft_data.marker_fd, FTRACE_ATTRS,
-				   "rtapp_attrs: event=uclamp util_min=%d",
-				   sched_data->util_min);
-		}
-		if (sched_data->util_max != -1) {
-			sa_params.sched_util_max = sched_data->util_max;
-			sa_params.sched_flags |= SCHED_FLAG_UTIL_CLAMP_MAX;
-			log_notice("[%d] setting util_max=%d",
-				   data->ind, sched_data->util_max);
-			log_ftrace(ft_data.marker_fd, FTRACE_ATTRS,
-				   "rtapp_attrs: event=uclamp util_max=%d",
-				   sched_data->util_max);
-		}
-
-		/*
-		 * We can't change clamp values without valid policy and
-		 * priority values.
-		 */
-		sa_params.sched_policy = sched_data->policy;
-		sa_params.sched_priority = sched_data->prio;
-
-		ret = sched_setattr(tid, &sa_params, flags);
-		if (ret != 0) {
-			log_critical("[%d] sched_setattr returned %d",
-				     data->ind, ret);
-			perror("sched_setattr: failed to set uclamp value(s)");
-			exit(EXIT_FAILURE);
-		}
-	}
-
+	sched_data->policy = __sched_policy(data, sched_data);
```

Comment:
The inconvenience with this approach, compared to bringing a new `sched_data` member for holding "same" , is that the `sched_data` is modified.

In the case of rt-app, looping over few phases, the first iteration will overwrite the `sched_data->policy`, and we'll loose the information "same" for the second iteration.

Is it still ok to pursue with this approach ?

Reply:
I could alternatively bring a new member into `thread_data_t` instead of `sched_data_t` ?

Reply:
Setting sched_data->policy with policy of the curr_sched_data->policy if current value is same is the current behavior and works just fine. Then compare sched_data->policy with curr_sched_data->policy to know if you need to change the policy instead of if (sched_data->policy != same) which will work only for the 1st time we run the phase

Reply:
The problem I have is with the following setup:

```
phase_0 {
    "policy": SCHED_OTHER
}
phase_1 {
    "priority": -10
}
```

```
# First loop
## Phase 0

$ setting scheduler SCHED_OTHER priority 0
# call pthread_setschedparam
# call setpriority

## Phase 1
$ setting scheduler SCHED_OTHER priority -10
# sched_data->policy == same .. skip pthread_setschedparam
# call setpriority to set -10.
# sched_data->policy = SCHED_OTHER

# Second loop

## Phase 0
...

## Phase 1
$ setting scheduler SCHED_OTHER priority -10
# call pthread_setschedparam. ... as now sched_data->policy == SCHED_OTHER
# call setpriority
```